platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,438 @@
+| Designers          | Richard Bair, Jasper Potts, Oleg Mazurov, Austin Littley |
+| Functional Impacts | Consensus Node                                           |
+
+---
+
+## Purpose and Context
+
+Much of the motivation for this design can come down to paying down technical debt and simplifying the overall design.
+While the current design is full of amazing high quality solutions to various problems, it is also more complex than
+necessary, leading to hard-to-find or predict bugs, performance problems, or liveness (stability) issues while under
+load. In addition, the separation of the "platform" and "services" in the consensus node is somewhat arbitrary, based
+on an abstract concept of what the "platform" was long before Hedera was even created. This work is also necessary to
+prepare for more autonomous node operation, and community nodes.
+
+Principally:
+1. This design defines several high-level modules, made up of internal "components". Whereas the current implementation
+   only has a logical grouping of components into systems, this design has concrete modules providing a strong
+   modularization and isolation with strict contracts between modules, leading to an overall simpler to understand
+   system, less prone to "spaghetti".
+2. Each module has very well-defined inputs and outputs.
+3. Assumptions and requirements that led to tight coupling between modules have been revisited, and where possible,
+   eliminated.
+4. The concepts of "platform" vs. "services" is less important in this design than the difference between "consensus"
+   and "execution". The Consensus module takes transactions and produces rounds. Everything required to make that happen
+   (gossip, event validation, hashgraph, event creation, etc.) is part of the Consensus module. It is a library, and
+   instances of the classes and interfaces within this library are created and managed by the Execution module. The
+   Consensus module has no state, no main method, and minimal dependencies.
+5. The Execution module is a mixture of what we have called "services" and some parts of "platform". The services code
+   has a `HandleWorkflow` and the platform has a number of classes that handle "post-consensus" responsibilities such
+   as managing the merkle tree, making fast copies of it, hashing it, etc. These responsibilities will be merged into
+   the `HandleWorkflow`, dramatically simplifying the interface and boundary between "consensus"/"platform" and
+   "execution"/"services".
+6. Backpressure, or dealing with a system under stress, will be radically redesigned based on
+   [PID](https://en.wikipedia.org/wiki/Proportional%E2%80%93integral%E2%80%93derivative_controller) controller logic,
+   and known as "dynamic throttles". The key concept is that we primarily throttle using information from _the entire
+   network_, rather than throttling only based on information available to the Consensus module like we do today.
```

Comment:
Dynamically throttling incoming transactions at the global level is a very important feature that the system is currently lacking, and so this is a very good thing to do. However, I do not think it replaces the need for individual nodes to apply internal back pressure based on local conditions. A system like this needs both global and local back pressure in order to be robust.

Suppose an honest node is 1% slower than its peers, and begins to be unable to keep up. Most nodes are happy with the workload, and so the global consensus is that the rate of transactions should not be decreased. (Any protocol that allows fewer than 1/3 of nodes by weight to apply global back pressure should be considered an attack vector given the security model).

Without local back pressure, that node may will thrash, with the eventual outcome being an OOM. It's much better if that node can gracefully pump the brakes. This allows the node to rejoin the network once the performance issues is resolved or the transactional workload changes.